ENH: add implementation of solver for the maximum flow problem

fuglede opened the pull request:
Following [@rgommers' reply](https://mail.python.org/pipermail/scipy-dev/2019-July/023601.html) to the mailing list, this includes an implementation of the [Edmonds--Karp algorithm](https://en.wikipedia.org/wiki/Edmonds%E2%80%93Karp_algorithm) for solving the [maximum flow problem](https://en.wikipedia.org/wiki/Maximum_flow_problem); this fits the bill of being a well-known algorithm, it beats e.g. NetworkX on performance, while probably being the simplest algorithm that solves the problem (but not the one with the best complexity guarantees) and it features in introductory material on graph theory (CLRS devotes a full chapter to the topic). What follows are some general remarks on the contents that didn't fit as code comments.

## Design considerations

For the most part, I've followed the lead from the rest of `scipy.sparse.csgraph` and attempted to stick to an API that's similar to what we already have; for convenience, [here's a reference](https://14398-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.sparse.csgraph.maximum_flow.html#scipy.sparse.csgraph.maximum_flow) to the documentation built for the new function.

For now, the PR includes only the Edmonds--Karp algorithm, but there are [several other solvers](https://en.wikipedia.org/wiki/Maximum_flow_problem#Solutions) that it would be interesting to include as well; a natural candidate would be the push-relabel algorithm whose complexity improves on that of Edmonds--Karp while still being simple to implement (and thus hopefully maintain). Besides the theoretical improvement, it also performs well in practice.

As such, there are some open questions about the API: I figured `scipy.sparse.csgraph.maximum_flow` would be a natural entry point for all of them, and to make sure that we leave open the option to include additional outputs, this one produces a `MaxFlowResult`, inspired by `optimize.OptimizeResult` but implemented as what's effectively just a `dataclass`, and which can be extended as necessary. This part could use additional scrutiny during review.


## Correctness

Besides including a small collection of unit tests, I ran a sanity check against NetworkX, comparing the results of the algorithm at hand with its `nx.flow.maximum_flow_value` for various collections of inputs, e.g.

```python
n = 1000
density = .1
for k in range(100):
    m = (scipy.sparse.rand(n, n, density=density, format='csr',
                           random_state=k)*100).astype(np.int32)
    G = nx.from_numpy_matrix(m.toarray(), create_using=nx.DiGraph())
    nx_max_flow = nx.algorithms.flow.maximum_flow_value(G, 0, n-1, capacity='weight')
    scipy_max_flow = maximum_flow(m, 0, n-1).flow_value
    assert nx_max_flow == scipy_max_flow
```


## Performance

A few words on performance and bottlenecks in the current implementation: While the Edmonds--Karp algorithm itself plays very well with CSR formatted input, it comes with the requirement that for any edge in the input directed graph, the reversed edge must also belong to the graph. We probably don't want to require this in the public API, so some fiddling is necessary to have this set up. Luckily, a combination of `scipy.sparse` helpers and various NumPy search functions makes it possible to vectorize the entire process; this is the content of `_add_reverse_edges` and `_make_pointers`. Together, these end up taking some 40% of the entire execution time, with most of the time spent in `tocoo` and `tocsr`; this seems like a lot, and maybe we could improve performance at the cost of memory efficiency, but the benefit is that we keep memory consumption at O(*E*) throughout, and the resulting memory locality probably saves us an equal amount in `_edmonds_karp` itself; indeed most of the time there is spent doing lookups. Moreover, this seemed more in the spirit of `scipy.*sparse*`.

As a simple benchmark, I compared some of the examples above to what's in NetworkX; we beat its Edmonds--Karp and preflow-push implementations by a good deal; how much depends on the characteristics of the input graph, but this should at the very least give a hint about the sort of improvements we should expect:

```python
In [3]: n = 1000
   ...: density = .1
   ...: m = (scipy.sparse.rand(n, n, density=density, format='csr', random_state=42)*100).astype(np.int32)
   ...: G = nx.from_numpy_matrix(m.toarray(), create_using=nx.DiGraph())
   
In [7]: %timeit nx.algorithms.flow.maximum_flow_value(G, 0, n - 1, capacity='weight', flow_func=edmonds_karp)
1.25 s ± 60 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit nx.algorithms.flow.maximum_flow_value(G, 0, n - 1, capacity='weight', flow_func=preflow_push)
1.19 s ± 111 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: %timeit scipy.sparse.csgraph.maximum_flow(m, 0, n - 1)
112 ms ± 12.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

And just to illustrate how this depends on the type of graph, here is a different family of examples in which vertices are lined up and only share edges with their 100 nearest neighbours. Here, we beat NetworkX's Edmonds--Karp by a large margin, and its preflow-push by a good amount as well.

```python
In [13]: n = 500
    ...: np.random.seed(42)
    ...: a = np.zeros((n, n), dtype=np.int32)
    ...: for k in range(n - 1):
    ...:     for j in range(-50, 50):
    ...:         if j != 0 and k + j >= 0 and k + j < n:
    ...:             a[k, k + j] = np.random.randint(1, 1000)
    ...: m = csr_matrix(a)
    ...: G = nx.from_numpy_matrix(a, create_using=nx.DiGraph())

In [14]: %timeit nx.algorithms.flow.maximum_flow_value(G, 0, n - 1, capacity='weight', flow_func=edmonds_karp)
5.11 s ± 308 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [15]: %timeit nx.algorithms.flow.maximum_flow_value(G, 0, n - 1, capacity='weight', flow_func=preflow_push)
291 ms ± 85.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [16]: %timeit scipy.sparse.csgraph.maximum_flow(m, 0, n - 1)
26.9 ms ± 357 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The results of the apv benchmark look as follows on a slowish laptop:

```
[100.00%] ··· ====== ============ =========== ===========
              --                   density
              ------ ------------------------------------
                n        0.1          0.3         0.5
              ====== ============ =========== ===========
               200    2.20±0.5ms    7.26±2ms    13.5±3ms
               500     19.8±5ms    61.6±10ms    169±8ms
               1500    220±20ms     960±70ms   2.30±0.1s
              ====== ============ =========== ===========
```

## Documentation check

One thing that's off in the [built documentation](https://14325-1460385-gh.circle-artifacts.com/0/html-scipyorg/sparse.csgraph.html) is that parameters are missing in the package overview:

![image](https://user-images.githubusercontent.com/6169306/62244409-ab07ba00-b3df-11e9-8f69-a661396c29d4.png)

I'm not sure why that is, or where to specify them, though?

rgommers commented on scipy/sparse/csgraph/bento.info:
```diff
@@ -5,6 +5,8 @@ Library:
     Extension: _min_spanning_tree
         Sources: _min_spanning_tree.c
+    Extension: _max_flow
+        Sources: _max_flow.c
```
oops, this is a leftover file from a build system we no longer support. cleaned up now

fuglede replied:
Gotcha; merged the change into this one and fixed the documentation issue.

rgommers commented:
> One thing that's off in the built documentation is that parameters are missing in the package overview:

In Cython code, we need to include the signature of the function at the top of the docstring. See for example the `csgraph.shortest_path` docstring.

rgommers commented:
@jakevdp perhaps you'd be interested in reviewing this PR?

fuglede commented:
If the prospective reviewer happens to be one with a surplus of time, they might also want to take a look at #10641 which is very similar to the one at hand.

fuglede commented:
Possible release note entry for this one:

> A solver for the maximum flow problem has been added as `scipy.sparse.csgraph.maximum_flow`

fuglede commented:
CI failure appears to be unrelated:
```
/home/circleci/repo/doc/source/tutorial/fft.rst:117: WARNING: Exception occurred in plotting fft-2
 from /home/circleci/repo/doc/source/tutorial/fft.rst:
Traceback (most recent call last):
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/pyparsing.py", line 5520, in oneOf
    return Regex("[%s]" % "".join(_collapseAndEscapeRegexRangeChars(sym) for sym in symbols)).setName(' | '.join(symbols))
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/pyparsing.py", line 5520, in <genexpr>
    return Regex("[%s]" % "".join(_collapseAndEscapeRegexRangeChars(sym) for sym in symbols)).setName(' | '.join(symbols))
NameError: name '_collapseAndEscapeRegexRangeChars' is not defined
```

fuglede commented:
This time around, the CI failures are happening in `TestLinprogSimplexDefault.test_disp` which is also unrelated to the PR.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
```diff
@@ -0,0 +1,395 @@
+
+    Returns
+    -------
+    res : MaximumFlowResult
```
This is a little unusual, but it does allow for extending the returns in the future. So I think it's fine. The one thing I'm wondering about is whether to put this class in the `csgraph` namespace - that pattern then adds two things to a namespace each time a single function is added. I'm inclined to not do that. In that case, the line below should use double backticks, because there won't be a doc page for `MaximumFlowResult` to link to.

fuglede replied:
Right, this might well be looking too far into the future. A place where having `MaximumFlowResult` would pay off is the following: Assume that the feature turns out to be popular; then a way to improve performance even further would be to expand the set of solver methods (low-hanging fruit would be [push–relabel](https://en.wikipedia.org/wiki/Push%E2%80%93relabel_maximum_flow_algorithm)); doing so, the collection of outputs from the general purpose `maximum_flow` entry point would depend on the solver, and collecting them in `MaximumFlowResult` would provide flexibility without the risk of breaking API changes.

I guess the class wouldn't be particularly useful outside of `maximum_flow` itself, so putting it in `csgraph` won't buy you much either? Perhaps other than consistency?

I'm not too familiar with Sphinx syntax; it [looks like](https://stackoverflow.com/a/22257273/5085211) the use of double backticks, but would I be right in understanding that single backticks are used whenever additional parsing, and double backticks are used everywhere else?

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
This comment is a little out of place, not immediately clear to me what is meant

fuglede replied:
Good catch. In an earlier version, the functionality of `_add_reverse_edges` was part of `_make_edge_pointers`; at some point, I extracted the method, making this comment outdated. Will remove.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
just a note on the `| |` (abs) symbols: those have a special meaning in Sphinx, and they shouldn't normally be used (it can break the doc build). I think it's not a problem here because it's in a private function so the docstring doesn't get parsed. But better to always enclose this in double backticks.

fuglede replied:
Gotcha. Then I think I'll opt to enclose them in `:math:` so that in case they do get parsed, they'll look good as well.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
References should go above Examples

fuglede replied:
The reference to CLRS ([2]) is used in the Examples section, but I guess that's fine.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
this should be spelled simply as `` `maximum_bipartite_matching` ``, Sphinx should be able to generate that link

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
```diff
@@ -0,0 +1,395 @@
+
+    def __init__(self, flow_value, residual):
+        self.flow_value = flow_value
+        self.residual = residual
```
If we are going to be returning a class instance, then adding a `__repr__` method would be good. This isn't too informative at the moment:
```
In [3]:  maximum_flow(graph, 0, 1)                                                             
Out[3]: <scipy.sparse.csgraph._flow.MaximumFlowResult at 0x11028ff28>
```

fuglede replied:
Since `MaximumFlowResult.residual` is typically going to be large, that one won't fit well into a `__repr__`, but I included one using `flow_value`:

```python
>>> maximum_flow(graph, 0, 1)
MaximumFlowResult with value of 5
```

Let me know what you think.

rgommers replied:
Indeed, that doesn't print nicely

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
The rest of the `csgraph` module uses `ITYPE` instead of `np.int32`, it would be good to be consistent. Doesn't change how the code works right now, but helps with maintenance. For context: https://github.com/scipy/scipy/pull/243#issuecomment-6233652

fuglede replied:
Makes sense. I updated this in the Cython functions, the wrapping pure-Python functions, but not in the example using `np.int`.

rgommers replied:
Looks good module my comment on the hardcoded max value

fuglede commented:
Thanks for the review @rgommers; looks like you guys have more than enough on your plates these days! I'll work my way through and mark the comments as resolved as I go along (unless you would rather that they are left as unresolved for a possible double-check).

fuglede commented:
This time it's `scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_immediate_updating` failing on Travis Python 3.6.

rgommers commented:
@fuglede if you see that a CI failure is unrelated to your PR, it's fine to just note that in a comment and leave things as they are. given the size of the test suite and CI build matrix, it's not uncommon to have occasional failures - we just ignore them and can merge a PR also if CI is not 100% green

rgommers commented:
> mark the comments as resolved as I go along (unless you would rather that they are left as unresolved for a possible double-check).

resolving all obviously fixed thing is fine (and helpful), thanks for asking.

rgommers commented:
the one CI failure is unrelated

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
Note:
```
>>> np.int is int                                                                              
True
```

You want to be using `np.int32`, `np.int64`, or the Python builtin `int`. Those aliases are in the numpy namespace for historical reasons and are best avoided

rgommers commented:
This wasn't completely obvious to me, so capturing here: in principle the problem is well-defined for floating point edge weights, however the algorithm used here isn't guaranteed to converge then. Non-integer min cost max flow seems to be an open research problem.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
We may forget to tweak this is `ITYPE` ever changes, so I suggest to use
```
ITYPE_MAX = np.iinfo(ITYPE).max
```
outside the main loop, and then use `ITYPE_MAX` here

rgommers commented:
Did some smoke testing, all behaved as expected. With a graph size of 1000, the solution is still found in under 1 ms.

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
No idea why tests on Python 3.5 are not failing, but this is invalid syntax. SciPy 1.4.0 will be the last release to support py35, so unfortunately this will need changing to `.format()` style strings

fuglede replied:
Hm, yes, this should be covered by `test_raises_when_source_is_out_of_bounds` and `test_raises_when_sink_is_out_of_bounds`, and at least the [Python 3.5 output](https://travis-ci.org/scipy/scipy/jobs/609932536#L17296) on Travis states that it's 

> adding 'scipy/sparse/csgraph/tests/test_flow.py'

so unless I've missed something about test discovery, it's not that the tests are skipped altogether.

rgommers replied:
I think I'm too lazy to try and reproduce that. I'm quite sure this can raise `SyntaxError` on py35, so let's just change it

fuglede replied:
Yep, changed. We shouldn't dig into the green tests?

rgommers commented on scipy/sparse/csgraph/_flow.pyx:
oops sorry my previous comment was wrong. All of these type declarations should be `ITYPE_t` rather than `int`.

fuglede replied:
But still retaining `ITYPE` where we had `np.int32`, right? If so, [`c153318`](https://github.com/scipy/scipy/pull/10566/commits/c153318aae5fa6832b1713acd90fc186a7d56169) should take care of that.

rgommers replied:
yes correct. using `ITYPE` where you'd use a numpy dtype, and `ITYPE_t` for Cython memoryviews or Cython `ndarray` objects (you don't use the latter, but other `csgraph` functions do).

fuglede replied:
Bonus question: This doesn't harm performance, right?

rgommers replied:
indeed, identical performance-wise

rgommers commented:
Once the last two open comments are addressed and CI finished, I think this is good to merge.

rgommers commented:
>  it beats e.g. NetworkX on performance

If you have that comparison, it might be nice to add the results to the description. Not sure if it's 1.5x or 100x faster:)

fuglede commented:
Thanks again; I went through the remaining comments.

Good point about floating point weights; we could add a note on that in the docstring?

Regarding the comparison with NetworkX, that's the one outlined in the "Performance" section in the PR description, so about 10x - 100x, depending a bit on the structure of the graph; not quite the ~250x we found in the companion PR (#10641) but still pretty solid. And there's room for improvement here; the alternative algorithms will beat the implementation on hands for many classes of graphs.

rgommers commented:
> Good point about floating point weights; we could add a note on that in the docstring?

Yes, that would be nice, thanks

rgommers commented:
> Regarding the comparison with NetworkX, that's the one outlined in the "Performance" section in the PR description,

`:facepalm:` sorry, I just didn't read far enough. PR description is great!

fuglede commented:
> Yes, that would be nice, thanks

Added a few words and rebased. Will be keeping an eye on CI but otherwise back to you.

> `:facepalm:` sorry, I just didn't read far enough. PR description is great!

No worries, and thanks!

rgommers commented:
All green, in it goes. Thanks @fuglede, very nice new feature!

rgommers commented:
Would you mind adding a release note to https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0?

fuglede commented:
Excellent, thanks!

> Would you mind adding a release note to https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0?

Sure, done.

fuglede commented:
I went ahead and made a few further benchmarks to see how we were doing compared to the implementations in [graph-tool](https://graph-tool.skewed.de/) which is based on the Boost Graph Library which aims to include high performance implementations of the solvers as well.

Using the same random matrices as in the original benchmark, but now with varying density, I get the following results:

Algorithm | Density: 0.1 | Density: 0.3 | Density: 0.5
--- | --- | --- | ---
nx.algorithms.flow.edmonds_karp | 1.07s | 3.2s | 6.39s
nx.algorithms.flow.preflow_push | 1.07s | 3.27s | 6.18s
nx.algorithms.flow.shortest_augmenting_path | 1.08s | 3.25s | 6.23s
gt.edmonds_karp_max_flow | 274ms | 2.84s | 10s
gt.push_relabel_max_flow | 71ms | 466ms | 1.42s
gt.boykov_kolmogorov_max_flow | 79ms | 463ms | 895ms
scipy.sparse.csgraph.maximum_flow | 64ms | 234ms | 580ms

So that's excellent.

The benchmark methodology was as follows: All of NetworkX, graph-tool, and SciPy play well with NumPy arrays, to ensure a level playing ground, each would take a NumPy array of shape (density*1000*1000, 3) corresponding to the data of a `coo_matrix`.

```python
def make_data(density):
    m = (rand(1000, 1000, density=density, format='coo', random_state=42)*100).astype(np.int32)
    return np.vstack([m.row, m.col, m.data]).T

data01 = make_data(0.1)
data03 = make_data(0.3)
data05 = make_data(0.5)
```

As such, each framework would calculate the value of a maximum flow as follows:

```
import graph_tool.all as gt
from scipy.sparse import coo_matrix, csr_matrix, rand
from scipy.sparse.csgraph import maximum_flow as scipy_mf
import networkx as nx


def graph_tool_max_flow(data, primitive):
    g = gt.Graph()
    cap = g.new_edge_property('int')
    eprops = [cap]
    g.add_edge_list(data, eprops=eprops)
    src, tgt = g.vertex(0), g.vertex(999)
    res = primitive(g, src, tgt, cap)
    res.a = cap.a - res.a
    return sum(res[e] for e in tgt.in_edges())


def scipy_max_flow(data):
    m = csr_matrix((data[:, 2], (data[:, 0], data[:, 1])))
    return scipy_mf(m, 0, 999).flow_value


def networkx_max_flow(data, primitive):
    m = coo_matrix((data[:, 2], (data[:, 0], data[:, 1])))
    G = nx.from_numpy_array(m.toarray(), create_using=nx.DiGraph())
    return nx.maximum_flow_value(G, 0, 999, capacity='weight', flow_func=primitive)
```

With this, examples of benchmarks become
```python
%timeit networkx_max_flow(data01, nx.algorithms.flow.shortest_augmenting_path)
%timeit graph_tool_max_flow(data03, gt.push_relabel_max_flow)
%timeit scipy_max_flow(data05)
```

Again, this would all look different for different graph structures, but it certainly seems like we're on the right track.

rgommers commented:
very nice, thanks @fuglede!
